### PR TITLE
plugin: add build rules for clients consuming an ES module set

### PR DESCRIPTION
**Pull request by psybuzz**
* Motivation for features / changes

To support clients who may need to slurp/compile a source set of ES modules, this PR modifies the plugin library to account for that case.

In particular, updating the plugin library now requires us to manually run some commands outlined in plugin_lib/BUILD's comment near "tb_pugin_lib".

### CODE REVIEW

File: tensorboard/components/experimental/plugin_lib/BUILD [hunk 1]
```diff
@@ -1,32 +1,43 @@
 package(default_visibility = ["//tensorboard:internal"])
 
 load("//tensorboard/defs:web.bzl", "tf_web_library")
+load("//tensorboard/defs:defs.bzl", "tf_ts_library", "tf_js_binary")
 
 licenses(["notice"])  # Apache 2.0
 
+# Serve for tests only.
```
Comment (stephanwlee) [inline tensorboard/components/experimental/plugin_lib/BUILD]: use the `testonly = 1` instead.

File: tensorboard/components/experimental/plugin_lib/BUILD [hunk 2]
```diff
@@ -1,32 +1,43 @@
...
-    visibility = ["//visibility:public"],
+)
+
+# To update the plugin lib for clients relying on a source set of ES6 Modules:
+# > bazel build tensorboard/components/experimental/plugin_lib:tb_plugin_lib
+# Then copy this folder
+# cp -r ./bazel-bin/tensorboard/components/experimental/plugin_lib/tb_plugin_lib.es6/tensorboard/components/experimental/plugin_lib/ tensorboard/components/experimental/plugin_lib/dist
+# into dist/
+tf_js_binary(
+    name = "tb_plugin_lib",
```
Comment (stephanwlee) [inline tensorboard/components/experimental/plugin_lib/BUILD]: if you are not going to commit the output of the binary, you probably don't need to have this target.

Instead, `bazel build tensorboard/components/experimental/plugin_lib:ts_plugin_lib` should be sufficient.

File: tensorboard/components/experimental/plugin_lib/dist/message.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
...dCBjYWxsYmFjayA9IHRoaXMubGlzdGVuZXJzLmdldCh0eXBlKSBhcyBNZXNzYWdlQ2FsbGJhY2s7XG4gICAgICB0cnkge1xuICAgICAgICBjb25zdCByZXN1bHQgPSBhd2FpdCBjYWxsYmFjayhwYXlsb2FkKTtcbiAgICAgICAgcmVwbHlQYXlsb2FkID0gcmVzdWx0O1xuICAgICAgfSBjYXRjaCAoZSkge1xuICAgICAgICByZXBseUVycm9yID0gZTtcbiAgICAgIH1cbiAgICB9XG4gICAgY29uc3QgcmVwbHlNZXNzYWdlOiBNZXNzYWdlID0ge1xuICAgICAgJ3R5cGUnOiB0eXBlLFxuICAgICAgJ2lkJzogaWQsXG4gICAgICAncGF5bG9hZCc6IHJlcGx5UGF5bG9hZCxcbiAgICAgICdlcnJvcic6IHJlcGx5RXJyb3IsXG4gICAgICAnaXNSZXBseSc6IHRydWUsXG4gICAgfTtcbiAgICB0aGlzLnBvc3RNZXNzYWdlKHJlcGx5TWVzc2FnZSk7XG4gIH1cblxuICBwcml2YXRlIHBvc3RNZXNzYWdlKG1lc3NhZ2U6IE1lc3NhZ2UpIHtcbiAgICB0aGlzLnBvcnQucG9zdE1lc3NhZ2UoSlNPTi5zdHJpbmdpZnkobWVzc2FnZSkpO1xuICB9XG5cbiAgc2VuZE1lc3NhZ2UodHlwZTogTWVzc2FnZVR5cGUsIHBheWxvYWQ6IFBheWxvYWRUeXBlKTogUHJvbWlzZTxQYXlsb2FkVHlwZT4ge1xuICAgIGNvbnN0IGlkID0gdGhpcy5pZCsrO1xuICAgIGNvbnN0IG1lc3NhZ2U6IE1lc3NhZ2UgPSB7J3R5cGUnOiB0eXBlLCAnaWQnOiBpZCwgJ3BheWxvYWQnOiBwYXlsb2FkLCAnZXJyb3InOiBudWxsLCAnaXNSZXBseSc6IGZhbHNlfTtcbiAgICB0aGlzLnBvc3RNZXNzYWdlKG1lc3NhZ2UpO1xuICAgIHJldHVybiBuZXcgUHJvbWlzZSgocmVzb2x2ZSwgcmVqZWN0KSA9PiB7XG4gICAgICB0aGlzLnJlc3BvbnNlV2FpdHMuc2V0KGlkLCB7cmVzb2x2ZSwgcmVqZWN0fSk7XG4gICAgfSk7XG4gIH1cbn1cbiJdfQ==
```
Comment (stephanwlee) [inline tensorboard/components/experimental/plugin_lib/dist/message.js]: this should be removed :\

File: tensorboard/components/experimental/plugin_lib/tf-plugin-lib.ts [hunk 4]
```diff
@@ -13,8 +11,7 @@
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
--->
-<link rel="import" href="../tf-plugin-util/message.html" />
-<script src="plugin-guest.js"></script>
+==============================================================================*/
+import * as _runs from './runs.js';
```
Comment (stephanwlee) [inline tensorboard/components/experimental/plugin_lib/tf-plugin-lib.ts]: is the file extension important?

File: tensorboard/components/experimental/plugin_lib/BUILD [hunk 5]
```diff
@@ -14,12 +14,13 @@ tf_web_library(
         "plugin-guest.ts",
         "tf-plugin-lib.ts",
     ],
+    testonly = 1,
     path = "/tf-plugin-lib",
     visibility = ["//visibility:public"],
```
Comment (stephanwlee) [inline tensorboard/components/experimental/plugin_lib/BUILD]: remove me and you are good!

File: tensorboard/components/experimental/plugin_lib/BUILD (hunk 1 above)
Comment (psybuzz) [reply tensorboard/components/experimental/plugin_lib/BUILD]: Done

File: tensorboard/components/experimental/plugin_lib/BUILD (hunk 2 above)
Comment (psybuzz) [reply tensorboard/components/experimental/plugin_lib/BUILD]: Actually, we need this, as the ts_plugin_lib only produces ES5 output.  Added a comment

File: tensorboard/components/experimental/plugin_lib/dist/message.js (hunk 3 above)
Comment (psybuzz) [reply tensorboard/components/experimental/plugin_lib/dist/message.js]: Done

File: tensorboard/components/experimental/plugin_lib/BUILD (hunk 5 above)
Comment (psybuzz) [reply tensorboard/components/experimental/plugin_lib/BUILD]: Done

File: tensorboard/components/experimental/plugin_lib/tf-plugin-lib.ts (hunk 4 above)
Comment (psybuzz) [reply tensorboard/components/experimental/plugin_lib/tf-plugin-lib.ts]: The way I've written the tests, I'm importing modules which need the ".js" to fetch properly.
I think there could be a way to make the tests work without the extension, but I would like to defer that to future investigation.

Comment (stephanwlee) [reply tensorboard/components/experimental/plugin_lib/tf-plugin-lib.ts]: That's fair. :+1:

**Comment by stephanwlee** [comment]
Closing this as it is superseded by #2747.

**Pull request closed without merging**